hpy/devel/src/runtime/ctx_type.c

```diff
@@ -400,8 +400,16 @@ ctx_Type_FromSpec(HPyContext ctx, HPyType_Spec *hpyspec,
         PyErr_NoMemory();
         return HPy_NULL;
     }
+    HPy_ssize_t basicsize;
+    if (hpyspec->legacy_headersize == 0) {
+        // XXX: How to handle alignment issues, if any?
+        basicsize = sizeof(struct {HPyObject_HEAD}) + hpyspec->basicsize;
+    }
+    else {
+        basicsize = hpyspec->basicsize;
+    }
```

Comment:
look at PR #50: we already had code to deal with alignment issues, but was killed when we (re)introduced `HPyObject_HEAD`.
Basically, the situation before PR 50 was similar at what you are aiming at now, with the difference that what was called `has_pyobject_head` is now `legacy_headersize`. Sorry, I should have pointed you to that PR earlier :(

Moreover:
1. you can use `sizeof(struct _HPyObject_head_s)`
2. we should check that `legacy_headersize == sizeof(struct _HPyObject_head_s)` and complain clearly if they don't match. This should catch most of the errors in which the user uses `offesetof` on a field which is not the first.
3. we should check that that if you use any of the `legacy_*` features, `legacy_headersize > 0`.
4. we could also introduce a small helper `is_legacy_spec()` which `return spec->legacy_headersize > 0`: it would make it much clearer what is the intent, IMHO, but that's just my 2 cents.

Reply:
Thanks for the link to #50!

I fear there might be a problem with `PyObject_HEAD_SIZE` though. `HPyGenericObject` ensures that the padding is as big or bigger than any possible padding, but what happens if the padding ends up being smaller because the user struct contains something like:

```c
struct {
    PyObject_HEAD
    char c;
} MyObject;
```

which would presumably have no padding at all between PyObject_HEAD and c.

Quite possibly I am missing something about how to use this.

Reply:
Ah, I think I see the solution. The case I described is the legacy case, so it is up to us to allocate things properly in the non-legacy case and to ensure that the member offsets, etc are set correctly. Objects might be slightly bigger on CPython than in the non-legacy case due to the extra padding.

Reply:
/cc @arigo 
I think that the original code comes from him (via cffi). At some point he convinced me that this was the correct thing to do, but now I no longer remember why 😅

Reply:
I think I convinced myself. :)